test/distributed/test_c10d.py

```diff
@@ -2989,6 +2989,134 @@ def test_param_layout_mismatch_error(self):
             with self.assertRaisesRegex(RuntimeError, ".* appears not to match strides of the same param in process 0"):
                 m_ddp = DistributedDataParallel(m, device_ids=[dev0], process_group=process_group)
 
+    @requires_gloo()
+    def test_ddp_comm_hook_future_passing(self):
+        """
+        This unit test verifies whether the Future object is passed properly.
+        The callback function creates a Future object and sets a value to it.
+        """
+        class test_ddp_comm_hook(nn.Module):
+            def __init__(self):
+                super(test_ddp_comm_hook, self).__init__()
+                self.t0 = Task()
+
+            def forward(self, x, rank):
+                return self.t0(x + rank)
+
+        def simple_hook(state: object, bucket: dist.GradBucket) -> torch.futures.Future:
+            fut = torch.futures.Future()
+            fut_result = []
+            # Outside then callback, set the result to all zeros.
+            fut.set_result([torch.zeros_like(t) for t in bucket.get_tensors()])
+
+            def fut_then(fut):
+                # Return all ones, something different from what fut has at its result.
+                return [torch.ones_like(t) for t in bucket.get_tensors()]
```

Comment:
Maybe return `fut.result() + torch.ones_like(t)` in order to exercise the return value of the original future?

Reply:
This is a good idea. I changed the unit test accordingly.